[Flang][OpenMP] Restructure recursive lowering in `createBodyOfOp`

kparzysz opened the pull request:
This brings `createBodyOfOp` to its final intended form. First, input privatization is performed, then the recursive lowering takes place, and finally the output privatization (lastprivate) is done.

This enables fixing a known issue with infinite loops inside of an OpenMP region, and the fix is included in this patch.

Fixes https://github.com/llvm/llvm-project/issues/74348.

Recursive lowering [5/5]

kparzysz committed:
[Flang][OpenMP] Separate creation of work-sharing and SIMD loops, NFC

These two constructs were both handled in `genOMP` for loop constructs.
There is some shared code between the two, but there are also enough
differences to separate these two cases into individual functions.

The shared code may be placed into a helper function later if needed.

Recursive lowering [1/5]

kparzysz committed:
[Flang][OpenMP] Restructure recursive lowering in `createBodyOfOp`

This brings `createBodyOfOp` to its final intended form. First, input
privatization is performed, then the recursive lowering takes place,
and finally the output privatization (lastprivate) is done.

This enables fixing a known issue with infinite loops inside of an
OpenMP region, and the fix is included in this patch.

Fixes https://github.com/llvm/llvm-project/issues/74348.

Recursive lowering [5/5]

kparzysz committed:
[Flang][OpenMP] Push genEval closer to leaf lowering functions

Recursive lowering [4/5]

kparzysz committed:
[Flang][OpenMP] Push genEval calls to individual operations, NFC

Introduce `genNestedEvaluations` that will lower all evaluations
nested in the given, accouting for a potential COLLAPSE directive.

Recursive lowering [2/5]

kparzysz committed:
[Flang][OpenMP] Handle SECTION construct from within SECTIONS

Introduce `createSectionOp`, invoke it from the SECTIONS construct for
each nested SECTION construct. This makes it unnecessary to embed
OpenMPSectionConstruct inside of OpenMPSectionConstruct anymore.

Recursive lowering [3/5]

skatrak commented on flang/test/Lower/OpenMP/infinite-loop-in-construct.f90:
```diff
@@ -0,0 +1,19 @@
+! See https://github.com/llvm/llvm-project/issues/74348
+
+! CHECK-LABEL: func.func @_QPsb
+! CHECK: omp.parallel
```
I think it would be a bit better to also check the general structure inside of omp.parallel. That is, checking the blocks and terminators that are created.

kparzysz replied:
That's done.

skatrak reviewed (commented):
Thanks Krzysztof! This looks ok to me, but I suggest waiting for approval from a second reviewer.

kiranchandramohan commented:
@tblah mentions that this patch fixes the unstructured code issue in worksharing loop with collapse. https://github.com/llvm/llvm-project/pull/77329.

kparzysz committed:
Merge branch 'main' into users/kparzysz/spr/a04-geneval-in-leafs

kparzysz committed:
Add parameter description for genNested

kparzysz committed:
Merge branch 'users/kparzysz/spr/a04-geneval-in-leafs' into users/kparzysz/spr/a05-complete-createBodyOfOp

kparzysz committed:
Add more detailed checks to testcase

kparzysz committed:
Add testcase from https://github.com/llvm/llvm-project/pull/77329

kparzysz committed:
Trigger new build

kparzysz committed:
Merge branch 'main' into users/kparzysz/spr/a05-complete-createBodyOfOp

kiranchandramohan commented on flang/lib/Lower/OpenMP.cpp:
```diff
@@ -2223,37 +2247,64 @@ static void createBodyOfOp(
+    firOpBuilder.setInsertionPointToEnd(exitBlock);
+    auto *term = Fortran::lower::genOpenMPTerminator(firOpBuilder,
+                                                     op.getOperation(), loc);
+    // Only insert lastprivate code when there actually is an exit block.
+    // Such a block may not exist if the nested code produced an infinite
+    // loop (this may not make sense in production code, but a user could
+    // write that and we should handle it).
```
Is it guaranteed that it is only when there is an infinite loop that no exit block exists?
Also who/what inserts the terminator for this case (when there is infinite loop)?

kparzysz replied:
We rely on the `genEval` for the nested code to do its thing.  If we consider the nested code (PFT/Evaluation) on its own, it will have a single entry point, and zero or more exit points.  If it creates multiple blocks, it will need to create its own branches to facilitate the control flow.  After `genEval` we remove the temporary terminator, then we look at all the blocks created during `genEval`.  We know that everything that is present inside of these blocks was created for the nested code (and not the enclosing construct we're lowering).  Terminators for our op haven't been created yet, and they can only placed in blocks that don't yet have a terminator.  If every block has a terminator then there is no way to leave the region by means of reaching the end of it (that is aside from function calls).

Based on that, the logical thing to do would be to insert a terminator into all non-terminated blocks.  This would work even if there is always at most one such block.

The assertion that there can only be one comes mostly from two things:
- the requirement from the OpenMP standard that a structured block exits at the bottom of the program text, and
- the current lowering seems to provide a single FIR location that corresponds to the "bottom".

If we can't be sure that the assertion is correct, we should be able to remove it, and process all exiting blocks.

kparzysz replied:
What I'm going to do is to create a unique exiting block:  create a new block, and for each yet-unterminated block, will insert a branch to the new one.  Then insert the OMP terminator into it, and process the lastprivate in there.

This will get rid of the assertion.

kiranchandramohan replied:
OK. Makes sense.

kparzysz committed:
Replace assertion with constructing unique block

If there are multiple exiting blocks, create a new empty block,
and insert branches into it from all the exiting blocks. This
will create a unique exiting block for inserting OMP terminator,
and lastprivate processing.

kiranchandramohan commented on flang/lib/Lower/OpenMP.cpp:
```diff
@@ -2223,37 +2222,100 @@ static void createBodyOfOp(
+    if (exits.empty())
+      return nullptr;
+    // If there already is a unique exiting block, do not create another one.
+    // Additionally, some ops (e.g. omp.sections) require onlt 1 block in
```
```suggestion
    // Additionally, some ops (e.g. omp.sections) require only 1 block in
```

kiranchandramohan reviewed (approved):
LGTM.

kparzysz committed:
Update flang/lib/Lower/OpenMP.cpp

Fix typo in comment

Co-authored-by: Kiran Chandramohan <kiran.chandramohan@arm.com>
